crates/uv-distribution/src/pyproject_mut.rs

```diff
@@ -0,0 +1,156 @@
+use std::fmt;
+use std::str::FromStr;
+
+use thiserror::Error;
+use toml_edit::{Array, DocumentMut, Item, RawString, TomlError, Value};
+
+use pep508_rs::Requirement;
+use pypi_types::{LenientRequirement, VerbatimParsedUrl};
+
+use crate::pyproject::PyProjectToml;
+
+/// Raw and mutable representation of a `pyproject.toml`.
+///
+/// This is useful for operations that require editing an existing `pyproject.toml` while
+/// preserving comments and other structure, such as `uv add` and `uv remove`.
+pub struct PyProjectTomlMut {
+    doc: DocumentMut,
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Failed to parse `pyproject.toml`")]
+    Parse(#[from] Box<TomlError>),
+    #[error("Dependencies in pyproject.toml are malformed")]
+    MalformedDependencies,
+}
+
+impl PyProjectTomlMut {
+    /// Initialize a `PyProjectTomlMut` from a `PyProjectToml`.
+    pub fn from_toml(pyproject: &PyProjectToml) -> Result<Self, Error> {
+        Ok(Self {
+            doc: pyproject.raw.parse().map_err(Box::new)?,
+        })
+    }
+
+    /// Adds a dependency.
+    pub fn add_dependency(&mut self, req: &Requirement) -> Result<(), Error> {
+        let deps = &mut self.doc["project"]["dependencies"];
+        if deps.is_none() {
+            *deps = Item::Value(Value::Array(Array::new()));
+        }
+        let deps = deps.as_array_mut().ok_or(Error::MalformedDependencies)?;
+
+        // Try to find a matching dependency.
+        let mut to_replace = None;
+        for (i, dep) in deps.iter().enumerate() {
+            if let Some(dep) = dep.as_str().and_then(try_parse_requirement) {
+                if dep.name.as_ref().eq_ignore_ascii_case(req.name.as_ref()) {
```

Comment:
`eq_ignore_ascii_case` isn't quite right here. These should be `ProjectName` which are already normalized.